src/custom-element.js

```diff
@@ -668,21 +701,58 @@ function createBaseAmpElementProto(win) {
         }
         this.implementation_.layout_ = this.layout_;
         this.implementation_.firstAttachedCallback();
-        if (!isUpgraded(this)) {
-          // amp:attached is dispatched from the ElementStub class when it replayed
-          // the firstAttachedCallback call.
+        if (!this.isUpgraded()) {
+          // amp:attached is dispatched from the ElementStub class when it
+          // replayed the firstAttachedCallback call.
           this.dispatchCustomEvent('amp:stubbed');
         } else {
           this.dispatchCustomEvent('amp:attached');
         }
       } catch (e) {
         reportError(e, this);
       }
+
+      // It's important to have this flag set in the end to avoid
+      // `resources.add` called twice if upgrade happens immediately.
+      this.everAttached = true;
     }
     this.resources_.add(this);
   };
 
   /**
+   * Try to upgrade the element with the provided implementation.
+   * @param {!./base-element.BaseElement=} opt_impl
+   * @private @final @this {!Element}
+   */
+  ElementProto.tryUpgrade_ = function(opt_impl) {
+    const impl = opt_impl || this.implementation_;
+    dev.assert(!isStub(impl), 'Implementation must not be a stub');
+    if (this.upgradeState_ != UpgradeState.NOT_UPGRADED) {
+      // Already upgraded or in progress or failed.
+      return;
+    }
+
+    // The `upgradeCallback` only allows redirect once for the top-level
+    // non-stub class. We may allow nested upgrades later, but they will
+    // certainly be bad for performance.
+    this.upgradeState_ = UpgradeState.UPGRADE_IN_PROGRESS;
+    const res = impl.upgradeCallback();
+    if (!res) {
+      // Nothing returned: the current object is the upgraded version.
+      this.completeUpgrade_(impl);
+    } else if (typeof res.then == 'function') {
```

Comment:
Is there a reason to manualyl test for promise-like? Why not let `Promise` do that for us?

```js
Promise.resolve(res || impl).then((impl) => {
  this.completeUpgrade_(impl);
}, (reason) => {
  this.upgradeState_ = UpgradeState.UPGRADE_FAILED;
  rethrowAsync(reason);
});
```

This also removes [Zalgo](http://blog.izs.me/post/59142742143/designing-apis-for-asynchrony) from our code.